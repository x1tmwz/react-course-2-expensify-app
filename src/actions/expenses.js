
import database from '../firebase/firebase'


const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = 'unknow',
            note = '',
            amount = 0,
            createAt = 0
        } = expenseData;
        const expense = {description, note,amount,createAt};
        return database.ref('expenses').push(expense).then((ref)=>{
            dispatch(addExpense({
                id:ref.key,
                ...expense
            }))
        })
    };
}
const removeExpense = ({ id = "" } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
})
const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})
export { addExpense, removeExpense, editExpense };