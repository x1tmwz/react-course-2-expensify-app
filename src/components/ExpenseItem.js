import React from 'react';
import { Link } from 'react-router-dom';


export  const ExpenseItem = (props) => (
    <div>
        <Link to={`/edit/${props.id}`}>
            <h3>Expense : {props.description}</h3>
        </Link>
        <ul>
            {props.note && <li>note : {props.note}</li>}
            <li>amount: {props.amount}</li>
            <li>create at: {props.createAt}</li>
        </ul>
        
    </div>
)
export default ExpenseItem;