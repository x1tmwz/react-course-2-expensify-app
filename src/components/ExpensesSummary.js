import React from 'react';
import getTotalExpense from '../selectors/expense-total';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import numeral from 'numeral';
import { Link } from 'react-router-dom'


export const ExpensesSummary = ({ expenseCount, expenseTotal }) => {
    const expenseWord = expenseCount === 1 ? 'expense' : 'expenses'
    const totalAmountFormat = numeral(expenseTotal / 100).format('$0,0.00');
    return (
        <div className="page-header">
            <div className='content-container'>
                <h1 className="page-header__title">
                    Viewing {<span>{expenseCount}</span>} {expenseWord} totalling {<span>{totalAmountFormat}</span>}
                </h1>
                <div className="page-header__actions">
                    <Link to="/create" className='button'>Add Expense</Link>
                </div>
            </div>
        </div>

    );
}
const mapStateToProps = (state) => {
    const expenses = getVisibleExpenses(state.expenses, state.filters)
    return {
        expenseCount: expenses.length,
        expenseTotal: getTotalExpense(expenses)
    }

}

export default connect(mapStateToProps)(ExpensesSummary)



