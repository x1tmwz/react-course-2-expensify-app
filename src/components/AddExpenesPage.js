import React from 'react';
import { connect } from 'react-redux';
import ExpensForm from './ExpensForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpenesPage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push("/");
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpensForm
                        onSubmit={this.onSubmit}
                    />
                </div>

            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
})


export default connect(undefined, mapDispatchToProps)(AddExpenesPage);