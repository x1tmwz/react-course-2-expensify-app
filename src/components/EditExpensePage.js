import React from 'react';
import { connect } from 'react-redux';
import { startEditExpense, startRemoveExpnese } from '../actions/expenses';
import ExpensForm from './ExpensForm';

export class EditExpensePage extends React.Component {

    startRemoveExpnese = () => {
        this.props.startRemoveExpnese(this.props.expense.id);
        this.props.history.push("/");
    }
    startEditExpense = (expense) => {
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpensForm
                        expense={this.props.expense}
                        onSubmit={(expense) => { this.startEditExpense(expense) }}
                    />
                    <button className="button button--secondary" onClick={this.startRemoveExpnese}>Remove Expense</button>
                </div>

            </div>
        );
    }
}
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
};
const mapDispatchToProps = (dispatch) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpnese: (id) => dispatch(startRemoveExpnese({ id }))
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);