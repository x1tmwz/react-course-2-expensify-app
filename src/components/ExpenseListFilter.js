import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setFillter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilter extends React.Component {
    state = {
        calendarFocused: null,
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }
    onTextChange = (e) => {
        const text = e.target.value.trim();
        this.props.setFillter(text);
    };
    onSelectChange = (e) => {
        const sortBy = e.target.value;
        if (sortBy === 'amount') this.props.sortByAmount();
        if (sortBy === 'date') this.props.sortByDate();
    }
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input
                            type='text'
                            value={this.props.filters.text}
                            onChange={this.onTextChange}
                            className="text-input"
                            placeholder="Search expenses.."
                        />
                    </div>
                    <div className="input-group__item">
                        <select
                            onChange={this.onSelectChange}
                            value={this.props.filters.sortBy}
                            className="select"
                        >
                            <option value='amount'>sort by amount</option>
                            <option value='date'>sort by Date</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            displayFormat={"DD/MM/YYYY"}
                            numberOfMonths={1}
                            isOutsideRange={() => { return false }}
                            showClearDates={true}
                        />
                    </div>
                </div>
            </div>





        )
    }
}
const mapStateToProps = (state) => {
    return {
        filters: state.filters
    };
};
const mapDispatchToProps = (dispatch) => ({
    setFillter: (text) => dispatch(setFillter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))

})
export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter);