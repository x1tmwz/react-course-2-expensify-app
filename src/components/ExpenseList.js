import React from 'react';
import ExpenseItem from './ExpenseItem';
import getVisibleExpenses from '../selectors/expenses';
import { connect } from 'react-redux';


export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className='list-body'>
            {
                props.expenses.length === 0 ? (
                    <div>
                        <span className="list-item list-item--message">
                            No Expenses
                    </span>
                    </div>
                ) :
                    (props.expenses.map((expense) => (<ExpenseItem key={expense.id} {...expense} />)))
            }
        </div>
    </div>
);
const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);

