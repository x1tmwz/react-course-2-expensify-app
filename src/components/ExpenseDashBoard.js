import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';
import ExpensesSummary from './ExpensesSummary'

const ExpenseDashBoard = () => (
    <div>
        <ExpensesSummary />
        <ExpenseListFilter />
        <ExpenseList />
    </div>
);

export default ExpenseDashBoard;