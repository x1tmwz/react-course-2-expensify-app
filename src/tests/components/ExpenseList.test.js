import { shallow } from 'enzyme';
import React from 'react';
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses'


test('shuold render Expenses with data', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});
test('shuold render Expenses without data', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});
