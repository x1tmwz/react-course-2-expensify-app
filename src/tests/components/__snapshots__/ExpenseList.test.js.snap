// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shuold render Expenses with data 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <ExpenseItem
      amount={1}
      createAt={0}
      description="cat"
      id="0"
      key="0"
      note=""
    />
    <ExpenseItem
      amount={2}
      createAt={-345600000}
      description="coffe"
      id="1"
      key="1"
      note=""
    />
    <ExpenseItem
      amount={3}
      createAt={345600000}
      description="rent"
      id="2"
      key="2"
      note=""
    />
  </div>
</div>
`;

exports[`shuold render Expenses without data 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div>
      <span
        className="list-item list-item--message"
      >
        No Expenses
      </span>
    </div>
  </div>
</div>
`;
